tl-its-umich-edu/canvas-course-manager-next
# Handle manually removed Canvas integration case (#101, #184)

The PR aims to resolve #101 and #184.

To Do
- [x] Implement an interceptor for catching errors related to an invalid token, deleting the token (if necessary), and sending custom response error data (including `redirect: true`).
- [x] Redirect in client to root using `location.href` when `401` is encountered and `redirect: true` is present in the error body (see `handleErrors`).
- [x] Modify globals to send a boolean indicating whether the user has a canvas token
- [x] Redirect in client using `location.href` to `canvas/redirectOAuth` if the user doesn't have a Canvas token
- [x] Always send the user to `/` first, instead of redirecting directly to `canvas/redirectOAuth` from `LTIService`
- [x] Enforce 1:1 relationship between `user` and `canvas_token` at database level by using unique constraint (#184)
 
Recommended testing procedures
- Ensure brand new user is successfully routed through OAuth and can subsequently make requests.
- Test requests with using invalid Canvas token are properly redirected to OAuth flow
  1. Launch application as normal.
  2. Go to `profile/settings` in Canvas and delete all "Approved Integrations" entries for your API Developer Key.
  3. Perform an action in the UI that involves a request to Canvas, checking that you are redirected to Canvas OAuth page.
  4. Repeat steps ii and iii until all actions are tested.
- Test requests involving a token refresh of an invalid Canvas token are properly redirect to OAuth flow
  1. Launch application as normal.
  2. Go to `profile/settings` in Canvas and delete all "Approved Integrations" entries for your API Developer Key.
  3. Modify the `expires_at` value in the `canvas_token` record for your user in the database so it is in the past.
  4. Perform an action in the UI that involves a request to Canvas, checking that you are redirected to Canvas OAuth page.
  5. Repeat steps ii, iii, and iv until all actions are tested.
- Test migration with `up` and `down` and `up` again (see `README.md`).

Resources
- https://docs.nestjs.com/interceptors
- https://sequelize.org/master/class/lib/dialects/abstract/query-interface.js~QueryInterface.html
